OgreMain/src/OgreSceneManager.cpp

```diff
@@ -2203,7 +2203,48 @@ void SceneManager::renderSingleObject(Renderable* rend, const Pass* pass,
             {
                 pLightListToUse = &rendLightList;
             }
-            lightsLeft = 0;
+
+            if (isShadowTechniqueIntegrated() && mLightsDirtyCounter > 0)
+            {
+                size_t shadowTexIndex = mShadowRenderer.getShadowTexIndex(lightIndex);
+                unsigned short numShadowTextureLights = 0;
+
+                for (const auto it : *pLightListToUse)
+                {
+                    if (!it->getCastShadows())
+                        continue;
+
+                    // potentially need to update content_type shadow texunit
+                    // corresponding to this light
+                    size_t textureCountPerLight = mShadowRenderer.mShadowTextureCountPerType[it->getType()];
+                    size_t textureCounterPerLight = 0;
+
+                    for (size_t j = 0; j < textureCountPerLight && shadowTexIndex < mShadowRenderer.mShadowTextures.size(); ++j)
+                    {
+                        // link the numShadowTextureLights'th shadow texture unit
+                        ushort tuindex = pass->_getTextureUnitWithContentTypeIndex(TextureUnitState::CONTENT_SHADOW, numShadowTextureLights);
+                        if (tuindex > pass->getNumTextureUnitStates()) break;
+
+                        TextureUnitState *tu = pass->getTextureUnitState(tuindex);
+
+                        //pick up correct shadow texture index by light's frame index + texture counter per light
+                        const TexturePtr &shadowTex = mShadowRenderer.mShadowTextures[mShadowRenderer.mShadowTextureIndexLightList[it->_getIndexInFrame()] + textureCounterPerLight];
+                        tu->_setTexturePtr(shadowTex);
+
+                        Camera *cam = shadowTex->getBuffer()->getRenderTarget()->getViewport(0)->getCamera();
+                        tu->setProjectiveTexturing(!pass->hasVertexProgram(), cam);
+                        mAutoParamDataSource->setTextureProjector(cam, numShadowTextureLights);
```

Comment:
this part of the loop looks familiar. Can we merge it with the existing loop to reduce complexity?

Reply:
This have to be investigated. Loop goes completely different depending on iteratePerLigh bool